### ISSUE TITLE
Opening a Vector Geopackage in QGIS alters its file even when no changes are done by user

### ISSUE THREAD

**Issue by qgib** [qgis/QGIS#23991]
Author Name: **Johannes Kroeger** (Johannes Kroeger)
Original Redmine Issue: [16076](https://issues.qgis.org/issues/16076)
Affected QGIS version: 3.6.0
Redmine category:unknown

---
Opening a Vector Geopackage in QGIS alters its file. Simply loading the file is enough, you do not even need to load any actual data (visibly to the user).

```
$ wget http://www.geopackage.org/data/geonames_belgium.gpkg
2017-01-13 17:58:54 (1.28 MB/s) - ‘geonames_belgium.gpkg’ saved [11594752/11594752]

$ cp geonames_belgium.gpkg geonames_belgium.gpkg_original

$ md5sum geonames_belgium.gpkg*

4dacf68ae3a5f2c2fb55a35d062dc580  geonames_belgium.gpkg
4dacf68ae3a5f2c2fb55a35d062dc580  geonames_belgium.gpkg_original

$ ogrinfo -so geonames_belgium.gpkg
INFO: Open of `geonames_belgium.gpkg'
      using driver `GPKG' successful.
1: administrative (3D Measured Point)
2: hydrography (3D Measured Point)
3: leisure (3D Measured Point)
4: places (3D Measured Point)
5: roads (3D Measured Point)
6: spots (3D Measured Point)
7: terrain (3D Measured Point)
8: undersea (3D Measured Point)
9: vegetation (3D Measured Point)

$ md5sum geonames_belgium.gpkg*
4dacf68ae3a5f2c2fb55a35d062dc580  geonames_belgium.gpkg
4dacf68ae3a5f2c2fb55a35d062dc580  geonames_belgium.gpkg_original

hannes@hilmar:/mnt/earx/geodata$ qgis geonames_belgium.gpkg
Warning: loading of qgis translation failed [/usr/share/qgis/i18n//qgis_en_US]
Warning: loading of qt translation failed [/usr/share/qt/translations/qt_en_US]
Warning: QCss::Parser - Failed to load file  "/style.qss"

```

Here the layer chooser dialog appeared on which I clicked Cancel and quit QGIS.

```
$ md5sum geonames_belgium.gpkg*
99fad69efe058b1d308088b96d60e915  geonames_belgium.gpkg
4dacf68ae3a5f2c2fb55a35d062dc580  geonames_belgium.gpkg_original

``` 

QGIS changes only some stuff in the header:

```
$ hexdump -C geonames_belgium.gpkg > geonames_belgium.gpkg.hex
$ hexdump -C geonames_belgium.gpkg_original > geonames_belgium.gpkg_original.hex
$ diff geonames_belgium.gpkg*.hex
2c2
< 00000010  04 00 01 01 00 40 20 20  00 00 00 56 00 00 2c 3b  |.....@  ...V..,;|
---
> 00000010  04 00 01 01 00 40 20 20  00 00 00 54 00 00 2c 3b  |.....@  ...T..,;|
6,7c6,7
< 00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 56  |...............V|
< 00000060  00 2e 05 42 05 00 00 00  26 03 42 00 00 00 2c 3b  |...B....&.B...,;|
---
> 00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 54  |...............T|
> 00000060  00 2d e6 06 05 00 00 00  26 03 42 00 00 00 2c 3b  |.-......&.B...,;|

```

I could not reproduce this bug with the Raster GPKG https://portal.opengeospatial.org/files/63156 but with any other Vector GPKG I tried.

Please ensure that merely opening a file does never alter it.

**Comment by strk** [comment]
@rouault the link you provided is not valid anymore (https://lists.osgeo.org/pipermail/qgis-user/2017-January/038830.html) -- do you have any updated reference as to why gpkg files should have to be modified when intended use is read-only ?

**Comment by rouault** [comment]
Perhaps this is: https://lists.osgeo.org/pipermail/qgis-user/2017-January/036509.html
The thing is that if you open in read-only in non-WAL mode, and then something else does read-write in WAL mode, I'm not sure what will happen.

**Comment by strk** [comment]
I think we can support the non-WAL mode at least during "make check", which would fix some of the https://github.com/qgis/QGIS/issues/25830 problems. Can we opt for non-WAL mode via some env variable ?

**Comment by saberraz** [comment]
You can do it in QGIS too:
https://github.com/qgis/QGIS/issues/31044#issuecomment-517912023

**Comment by florisvdh** [comment]
Clearly there are advantages of WAL-mode as a safer way of editing GeoPackages. However it's unclear to me why this choice should invoke the problem enlisted here, first upon opening the data and also on closing, and on closing in case _no_ changes have been made to the file. Isn't this something QGIS controls, even while using WAL? The philosophy of WAL appears to be: collect all edits in the separate 'wal' file, meanwhile don't change the original GPKG file. IMO nothing should change to the original file, not on opening the GeoPackage, and not if the 'wal' file has size zero (which means 'no changes made' IIUC).

I confirm the problem is present in QGIS 3.20.1-Odense. Adding another example below.

The GeoPackage is changed two times: first when opening (wal and shm files are added), and another time when closing (wal and shm files disappear - wal was 0 bytes).

```bash
$ wget -q https://github.com/r-spatial/sf/raw/7e60ab32c07fff804139e763e3e4ed0afa419ca3/inst/gpkg/b_pump.gpkg
$ md5sum b_pump.gpkg 
18e20ba28fe1b85ed94f9d5debcb1771  b_pump.gpkg
$ 
$ qgis b_pump.gpkg &> /dev/null &  # opens file in QGIS
[1] 7103
$ 
$ ls -l b_pump.gpkg*
-rw-rw---- 1 floris floris 98304 aug 10 10:12 b_pump.gpkg
-rw-rw---- 1 floris floris 32768 aug 10 10:12 b_pump.gpkg-shm
-rw-rw---- 1 floris floris     0 aug 10 10:12 b_pump.gpkg-wal
$ md5sum b_pump.gpkg
ad8691c170982bb45ccf311c678fc5ae  b_pump.gpkg
$ 
$ # closing QGIS now
$ 
[1]+  Done                    qgis b_pump.gpkg &> /dev/null
$ 
$ ls -l b_pump.gpkg*
-rw-rw---- 1 floris floris 98304 aug 10 10:12 b_pump.gpkg
$ md5sum b_pump.gpkg
760c3a04af039577bfbbed574d11baa1  b_pump.gpkg
```

The `last_change` field of `gpkg_contents` table still has its original contents.

I have an interest in making and distributing binary reproducible data files, such as GeoPackage files; so checking integrity of a file copy (checksum calculation) is a part of that. For that, QGIS should indeed not change a data file just because the user has had a look at the file contents. Getting all my co-workers (let alone other scientists using the data) to put their local QGIS installation to a non-WAL mode is not that easy. Also it seems like overkill since it would drop the advantages of WAL in effectively editing a GeoPackage.

Thank you for considering this further.

**Comment by jratike80** [comment]
> Author Name: **Even Rouault** (@rouault)
> 
> What you observe is by design to avoid dead lock issues in concurrent reading/writing scenarios. See https://lists.osgeo.org/pipermail/qgis-user/2017-January/038830.html

This link does not work now (2022-01-31)

**Comment by pigreco** [comment]
QGIS 3.22.5.1 and QGIS 3.24.1.2 with GDAL 3.4.2 continues to modify the last modification date without any data change:

https://youtu.be/yJeMvBE67xw

**Comment by kannes** [comment]
Check https://github.com/qgis/QGIS/pull/47098#issuecomment-1077842392 please

**Comment by strk** [comment]
For the record: this is still the case for some files which were created by GDAL < 3.3.1 -- the world_map.gpkg is one of those, see https://github.com/qgis/QGIS/issues/48937#issuecomment-1152740185

**Comment by gotjoshua** [comment]
why is this issue closed? 

it still works like this:
the gpkg is altered
[> first on opening the data and also on closing in case _no_ changes have been made to the file.](https://github.com/qgis/QGIS/issues/23991#issuecomment-895834279)

**Comment by strk** [comment]
According to the record it was reported to be closed by commit https://github.com/qgis/QGIS/commit/fe410ec92051b5be1d73b5753f05a7cfdc004112 -- if this is the case my PR #49124 might confirm it.

The comment linked in previous comment ( https://github.com/qgis/QGIS/pull/47098#issuecomment-1077842392 ) says it takes building against GDAL-3.4.2 (compile time) to see the fix.

It looks to me that the Dockerfile used by CI when testing against QT5 (.docker/qgis3-qt5-build-deps.dockerfile ) is using whatever GDAL version is shipped with Ubuntu-22.04 which would be 3.4.1 so I'm afraid we can't see the effects of the fix.
